ocaml/testsuite/tests/typing-layouts-products/product_arrays.ml

```diff
@@ -0,0 +1,1911 @@
+type t1 = #(int * bool option) array
+type t2 = #(int * string) array
+type t3 = #(string * int * int option) array
+type t4 : immediate & value
+type t4a = t4 array
+type t5 : value & immediate
+type t5a = t5 array
+type t6 : value & immediate & value & immediate
+type t6a = t6 array
+|}]
+
+(*******************************************************)
+(* Test 2: Some allowed ignorable product array types. *)
+type t1 = #(int * int) array
+type t2 = #(int * float#) array
+type t3 = #(float# * int * int64# * bool) array
+
+type t4 : immediate & immediate
+type t4a = t4 array
+type t5 : immediate & float64
+type t5a = t5 array
+type t6 : bits64 & immediate & float64 & immediate
+type t6a = t6 array
+[%%expect{|
+type t1 = #(int * int) array
+type t2 = #(int * float#) array
+type t3 = #(float# * int * int64# * bool) array
+type t4 : immediate & immediate
+type t4a = t4 array
+type t5 : immediate & float64
+type t5a = t5 array
+type t6 : bits64 & immediate & float64 & immediate
+type t6a = t6 array
+|}]
+
+(******************************************************************************)
+(* Test 3: Some array types that are allowed even though you can't make them. *)
+type t1 = #(float# * string) array
+type t2 = #(string * int64#) array
+type t3 = #(string * int64# * int) array
+type t4 = #(int * int64# * string) array
+
+type t5 : value & float64
+type t5a = t5 array
+type t6 : bits64 & value
+type t6a = t6 array
+type t7 : value & bits64 & immediate
+type t7a = t7 array
+type t8 : immediate & bits64 & value
+type t8a = t8 array
+
+[%%expect{|
+type t1 = #(float# * string) array
+type t2 = #(string * int64#) array
+type t3 = #(string * int64# * int) array
+type t4 = #(int * int64# * string) array
+type t5 : value & float64
+type t5a = t5 array
+type t6 : bits64 & value
+type t6a = t6 array
+type t7 : value & bits64 & immediate
+type t7a = t7 array
+type t8 : immediate & bits64 & value
+type t8a = t8 array
+|}]
+
+(*****************************)
+(* Test 4: makearray_dynamic *)
+
+(* An array poly version works at valid product layouts. *)
+external[@layout_poly] make_vect : ('a : any_non_null) . int -> 'a -> 'a array =
+  "%makearray_dynamic"
+
+let f_scannable (x : #(int * float * string) array) = make_vect 42 x
+
+let f_ignorable (x : #(float# * int * int64# * bool) array) = make_vect 42 x
+[%%expect{|
+external make_vect : ('a : any_non_null). int -> 'a -> 'a array
+  = "%makearray_dynamic" [@@layout_poly]
+val f_scannable :
+  #(int * float * string) array -> #(int * float * string) array array =
+  <fun>
+val f_ignorable :
+  #(float# * int * int64# * bool) array ->
+  #(float# * int * int64# * bool) array array = <fun>
+|}]
+
+(* But not on the bad ones. *)
+let f_bad (x : #(string * float#) array) = make_vect 42 x
+[%%expect{|
+Line 1, characters 10-40:
+1 | let f_bad (x : #(string * float#) array) = make_vect 42 x
+              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Unboxed product array elements must be external or contain all gc
```

Comment:
We resolved this in person: In principle we could allow such a thing, but the implementation of `layout_poly` externals makes it tricky, and it doesn't seem worth complicating things to suppport since you can't make anything of these types anyway.